telco-examples/.release_generator/generator.py

```diff
@@ -15,75 +33,243 @@ def increase_indent(self, flow=False, indentless=False):
+            image_map[image_name] = full_image
+    return image_map
 
 # -----------------------------
 # Update logic
 # -----------------------------
-def update_versions_in_mgmt_file(mgmt, release_versions):
-    # Update Kubernetes version
-    if "kubernetes" in mgmt and "version" in mgmt["kubernetes"]:
-        mgmt["kubernetes"]["version"] = release_versions.get("__k8s_version__", mgmt["kubernetes"]["version"])
+def update_versions_in_mgmt_file(doc, release_versions, image_map):
+    if "kubernetes" in doc and "version" in doc["kubernetes"]:
+        doc["kubernetes"]["version"] = release_versions.get("__k8s_version__", doc["kubernetes"]["version"])
 
-    # Update Helm chart versions
-    charts = mgmt.get("kubernetes", {}).get("helm", {}).get("charts", [])
+    charts = doc.get("kubernetes", {}).get("helm", {}).get("charts", [])
     for chart in charts:
         name = chart.get("name")
         if name in release_versions:
             chart["version"] = release_versions[name]
 
-    return mgmt
+    # Walk through any keys and try to replace image values
+    def replace_images(obj):
+        if isinstance(obj, dict):
+            for k, v in obj.items():
+                if isinstance(v, str):
+                    match = re.match(r".*/([^:/]+):[^:]+", v)
+                    if match:
+                        image_name = match.group(1)
+                        if image_name in image_map:
+                            obj[k] = image_map[image_name]
+                else:
+                    replace_images(v)
+        elif isinstance(obj, list):
+            for item in obj:
+                replace_images(item)
+
+    replace_images(doc)
+
+# -----------------------------
+# Preprocess placeholders: ${...}
+# -----------------------------
+def preserve_placeholders(yaml_text):
+    placeholders = {}
+    def replacer(match):
+        key = match.group(1)
+        token = f"__PLACEHOLDER_{len(placeholders)}__"
+        placeholders[token] = f"${{{key}}}"
+        return token
+
+    cleaned = re.sub(r"\$\{([^}]+)\}", replacer, yaml_text)
+    return cleaned, placeholders
+
+def restore_placeholders(yaml_text, placeholders):
+    for token, original in placeholders.items():
+        yaml_text = yaml_text.replace(token, original)
+    return yaml_text
+
+# -----------------------------
+# Inline text block update for version substitution
+# -----------------------------
+def update_inline_versions_in_text(text, release_versions):
+    def replace_inline_block(match):
+        block = match.group(0)
+        lines = block.splitlines(keepends=True)
+
+        current_release = None
+        new_lines = []
+
+        for line in lines:
+            name_match = re.match(r"^\s*name:\s*(\S+)", line)
+            if name_match:
+                current_release = name_match.group(1)
+
+            version_match = re.match(r"^(\s*)version:\s*([^\n]+)", line)
+            if version_match and current_release in release_versions:
```

Comment:
This feels fragile, as having version field before name field would ruin everything